nutils/evaluable.py

```diff
@@ -1629,159 +1631,144 @@ class Multiply(Array):
 
     def __init__(self, funcs: types.frozenmultiset):
         assert isinstance(funcs, types.frozenmultiset), f'funcs={funcs!r}'
-        self.funcs = funcs
-        func1, func2 = funcs
-        assert equalshape(func1.shape, func2.shape) and func1.dtype == func2.dtype != bool, 'Multiply({}, {})'.format(func1, func2)
-        super().__init__(args=tuple(self.funcs), shape=func1.shape, dtype=func1.dtype)
+        func, *others = funcs
+        assert all(equalshape(f.shape, func.shape) and f.dtype == func.dtype for f in others)
+        self.funcs = tuple(funcs)
+        super().__init__(args=tuple(funcs), shape=func.shape, dtype=func.dtype)
 
     def _simplified(self):
-        func1, func2 = self.funcs
-        if func1._const_uniform == 1:
-            return func2
-        if func2._const_uniform == 1:
-            return func1
-        unaligned1, unaligned2, where = unalign(func1, func2)
-        if len(where) != self.ndim:
-            return align(unaligned1 * unaligned2, where, self.shape)
-        for axis1, axis2, *other in map(sorted, func1._diagonals or func2._diagonals):
-            return diagonalize(multiply(takediag(func1, axis1, axis2), takediag(func2, axis1, axis2)), axis1, axis2)
-        for i, parts in func1._inflations:
-            return util.sum(_inflate(f * _take(func2, dofmap, i), dofmap, self.shape[i], i) for dofmap, f in parts.items())
-        for i, parts in func2._inflations:
-            return util.sum(_inflate(_take(func1, dofmap, i) * f, dofmap, self.shape[i], i) for dofmap, f in parts.items())
-        return func1._multiply(func2) or func2._multiply(func1)
+        for j, fj in enumerate(self.funcs):
+            if isinstance(fj, Multiply):
+                return multiply(*self.funcs[:j], *fj.funcs, *self.funcs[j+1:])
+            if fj._const_uniform == 1:
+                return multiply(*self.funcs[:j], *self.funcs[j+1:])
+            for i, parts in fj._inflations:
+                return util.sum(_inflate(multiply(f, *(_take(fi, dofmap, i) for fi in self.funcs[:j] + self.funcs[j+1:])), dofmap, self.shape[i], i) for dofmap, f in parts.items())
+            for axis1, axis2, *other in map(sorted, fj._diagonals):
+                return diagonalize(multiply(*(takediag(f, axis1, axis2) for f in self.funcs)), axis1, axis2)
+            for i, fi in enumerate(self.funcs[:j]):
+                unaligned1, unaligned2, where = unalign(fi, fj)
```

Comment:
If we're outer-multiplying three arrays (e.g. `einsum('i,j,k->ijk', ...)`) then the result of this simplification depends on the order of the multiplication: the first two arguments are unaligned, multiplied and aligned, then multiplied with the third argument. This seems opposite to what the patch tries to achieve. Should this simplification not be part of stage two (`_optimized_for_numpy`)? We can anyhow unalign all arguments simultaneously.

Reply:
The `unalign` method identifies insertions that are common to the arguments, so this is not a `i,j,k` situation but rather the multiplication of two `InsertAxes` objects. Which means that this could probably equally reside in `InsertAxis._multiply` -- still not sure which is better. Note that the old `Multiply._simplified` featured the same simplification.

Reply:
I understand that this gives the same simplification as before, and that is a valid point to keep it like this.

To clarify, my question concerns the following situation:

```python
from nutils import evaluable
arg = lambda name, length: evaluable.Argument(name, (evaluable.asarray(length),))
abc = evaluable.einsum('i,j,k->ijk', arg('a', 2), arg('b', 3), arg('c', 4)).simplified
```

With this PR (and current master) the tree of `abc` (note: simplified!) is

```
%0 = Multiply; f:2,3,4
: %1 = InsertAxis; f:2,3,4
| : %2 = Multiply; f:2,3
| | : %3 = InsertAxis; f:2,3
| | | : Argument; a; f:2
| | | : 3
| | : %4 = Transpose; 1,0; f:2,3
| |   : %5 = InsertAxis; f:3,2
| |     : Argument; b; f:3
| |     : 2
| : 4
: %6 = Transpose; 1,2,0; f:2,3,4
  : %7 = InsertAxis; f:4,2,3
    : %8 = InsertAxis; f:4,2
    | : Argument; c; f:4
    | : 2
    : 3
```

The choice to group and unalign the first two items of the multiplication seems arbitrary to me. Why not the first and the third. Or the second and the third. I have no objection against applying unalign to all arguments at once.

For reference, this is the tree of `abc` without unalign simplification:

```
%0 = Multiply; f:2,3,4
: %1 = InsertAxis; f:2,3,4
| : %2 = InsertAxis; f:2,3
| | : Argument; a; f:2
| | : 3
| : 4
: %3 = Transpose; 1,0,2; f:2,3,4
| : %4 = InsertAxis; f:3,2,4
|   : %5 = InsertAxis; f:3,2
|   | : Argument; b; f:3
|   | : 2
|   : 4
: %6 = Transpose; 1,2,0; f:2,3,4
  : %7 = InsertAxis; f:4,2,3
    : %8 = InsertAxis; f:4,2
    | : Argument; c; f:4
    | : 2
    : 3
```

Reply:
In terms of flops the former is better though, even if the choice is arbitrary. The fact that simplify does not bring the function in a canonical form is unfortunate but not specific to this operation. To mention just one example: the two `InsertAxis` operations follows by a `Transpose` could be arbitrarily swapped. The same applies to pretty much any sequence of operations of the same type. Not to say that the problem is not worth exploring, but I do not think the present PR is the appropriate opportunity. As for its intentions I have added some context in the description.

Reply:
But simplify is the flops reducing step; optimize merely brings it in a form that is more efficient to evaluate in a given backend (just numpy for now). This is why take is pushed up and inflate pulled down during simplify, and in this case inserted axes are merged where possible. I think it's important to stay close to what each step is aiming to accomplish and not blur any lines, given the complexity of the framework. Ultimately we will want to make simplify more powerful than the current greedy algorithm and then we will benefit from maintaining a separation of tasks.

Reply:
Consider this example: `a`, `b` and `c` are 3D arrays, each with two inserted axes (think of meshgrid). With this PR simplifying `a + a * b * c` gives `a + (a * b) * c` and simplifying `a + c * b * a` gives `a * (b * c + 1)`  (transposes and inserted axes omitted). The latter is made possible by the PR. Example code:

```python
from nutils.evaluable import align, asarray, Argument
shape = tuple(map(asarray, (2, 3, 4)))
a = align(Argument('a', shape[0:1]), (0,), shape)
b = align(Argument('b', shape[1:2]), (1,), shape)
c = align(Argument('c', shape[2:3]), (2,), shape)
f1 = a + a * b * c
f2 = a + c * b * a
print(f1.simplified.asciitree())
print(f2.simplified.asciitree())
```

I don't know which of the two is better and there will always be situations where the order of the operations affect the simplification, but this one is a avoidable. For instance, we could add an `_unmultiply` property that undoes the unalign simplification.

As I see it the objective of this PR is twofold:

* Enable more simplifications by providing flattened adds and multiplies.
* More efficient `{Add,Multiply}.evalf` by not having to allocate intermediate arrays.

As it stands now they are in an arms race.